src/sc/sidechainTxsCommitmentBuilder.cpp

```diff
@@ -0,0 +1,203 @@
+#include <primitives/certificate.h>
+
+#ifdef BITCOIN_TX
+void SidechainTxsCommitmentBuilder::add(const CTransaction& tx) { return; }
+void SidechainTxsCommitmentBuilder::add(const CScCertificate& cert) { return; }
+uint256 SidechainTxsCommitmentBuilder::getCommitment() { return uint256(); }
+#else
+void SidechainTxsCommitmentBuilder::add(const CTransaction& tx)
+{
+    if (!tx.IsScVersion())
+        return;
+
+    LogPrint("sc", "%s():%d -getting leaves for vsc out\n", __func__, __LINE__);
+    for(unsigned int scIdx = 0; scIdx < tx.GetVscCcOut().size(); ++scIdx)
+    {
+        uint256 scId = tx.GetVscCcOut().at(scIdx).GetScId();
+        sScIds.insert(scId);
+
+        std::vector<field_t*>& vec = mScMerkleTreeLeavesFt[scId]; //create or pick entry for scId
+        field_t* pField = mapScTxToField(tx.GetVscCcOut().at(scIdx).GetHash(), tx.GetHash(), scIdx);
+        vec.push_back(pField);
+    }
+
+    unsigned int fwdBaseMapIdx = tx.GetVscCcOut().size();
+    LogPrint("sc", "%s():%d -getting leaves for vft out\n", __func__, __LINE__);
+    for(unsigned int fwdIdx = 0; fwdIdx < tx.GetVftCcOut().size(); ++fwdIdx)
+    {
+        uint256 scId = tx.GetVftCcOut().at(fwdIdx).GetScId();
+        sScIds.insert(scId);
+
+        std::vector<field_t*>& vec = mScMerkleTreeLeavesFt[scId]; //create or pick entry for scId
+        field_t* pField = mapScTxToField(tx.GetVftCcOut().at(fwdIdx).GetHash(), tx.GetHash(), fwdBaseMapIdx + fwdIdx);
+        vec.push_back(pField);
+    }
+}
+
+void SidechainTxsCommitmentBuilder::add(const CScCertificate& cert)
+{
+    uint256 scId = cert.GetScId();
+    sScIds.insert(scId);
+    mScCerts[scId] = mapCertToField(cert.GetHash());
+}
+
+uint256 SidechainTxsCommitmentBuilder::getCommitment()
+{
+    std::vector<field_t*> vSortedScLeaves;
+
+    // set of scid is ordered
+    for (const auto& scid : sScIds)
+    {
+        field_t* ftRootField = nullptr;
+        auto itFt = mScMerkleTreeLeavesFt.find(scid);
+        if (itFt != mScMerkleTreeLeavesFt.end() )
+        {
+            ftRootField = merkleTreeRootOf(itFt->second);
+        }
+
+        field_t* btrRootField = nullptr;
+        auto itBtr = mScMerkleTreeLeavesBtr.find(scid);
+        if (itBtr != mScMerkleTreeLeavesBtr.end() )
+        {
+            btrRootField = merkleTreeRootOf(itBtr->second);
+        }
+
+        field_t* certRootField = nullptr;
+        auto itCert = mScCerts.find(scid);
+        if (itCert != mScCerts.end() )
+        {
+            certRootField = itCert->second;
+        }
+
+        field_t* scIdField = mapCertToField(scid);
+
+        std::vector<field_t*> singleSidechainComponents = {ftRootField, btrRootField, certRootField, scIdField};
+        field_t * sidechainTreeRoot = merkleTreeRootOf(singleSidechainComponents);
+
+        vSortedScLeaves.push_back(sidechainTreeRoot);
+    }
+
+    field_t * finalTreeRoot = merkleTreeRootOf(vSortedScLeaves);
+    uint256 res = mapFieldToHash(finalTreeRoot);
```

Comment:
Seems, that we need to return exactly poseidon hash, which will lead to the changes in the Block structure.
@DanieleDiBenedetto In the sidechain we have a problem now: `MerklePath.validate(...)` expects the root hash (poseidon one) as one of the arguments, but, we have only the sha hash of that root hash.

Temporary this can be solved by introducing `MerklePath.apply(leaf)` method that will return poseidon root hash. Then it can be hashed again with sha and compared with the on in MC block header. But, seems, we need to avoid sha at all.